core/src/main/scala/kafka/log/LogCleaner.scala

```diff
@@ -658,20 +673,22 @@ private[log] class Cleaner(val id: Int,
           }
         }
 
-        if (batch.hasProducerId && isBatchLastRecordOfProducer)
-          BatchRetention.RETAIN_EMPTY
-        else if (discardBatchRecords)
-          BatchRetention.DELETE
-        else
-          BatchRetention.DELETE_EMPTY
+        val batchRetention: BatchRetention =
+          if (batch.hasProducerId && isBatchLastRecordOfProducer)
+            BatchRetention.RETAIN_EMPTY
+          else if (discardBatchRecords)
+            BatchRetention.DELETE
+          else
+            BatchRetention.DELETE_EMPTY
+        new RecordFilter.BatchRetentionResult(batchRetention, canDiscardBatch)
```

Comment:
It seems that containsMarkerForEmptyTxn should only be set to canDiscardBatch if this batch is a control batch?

Reply:
hmm yeah I think you are right. I'll change to `canDiscardBatch && batch.isControlBatch`